Turn on transfer floating point round-off decisions diagnosis by default

GiudGiud opened the pull request:
refs #24226

This is going to have quite a few failures.
I ll try to tell the users to turn it off if not needed

GiudGiud committed:
Turn on transfer diagnosis by default, refs #24226

GiudGiud committed:
Manage verbosity to tell the user to tell the user to turn off diagnosis when things are well
refs #24226

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1491448/) on 4215e7d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24196/site)

This comment will be updated on new commits.

GiudGiud committed:
Adapt tests to new default for searching for value conflicts
This flags tests that may become brittle in the future, as they rely on floating point roundoff

GiudGiud committed:
Adapt to new multiapp terminology
refs #20124

roystgnr commented:
Do you have an approximate notion of how expensive it is to leave this on?

Is there a risk of this biting people in moving mesh simulations?  Like, they do a few timesteps, they're comfortable that they haven't got any transfer ambiguity, then they start a whole run and the next thing you know timestep 200 dies because a deforming mesh put a node into the wrong place.  I'm guessing "no", both because deformation will never go *exactly* to the wrong place and because "just turn off the flag and restart afterward" isn't too heinous a workaround; just trying to brainstorm possible failure cases.

GiudGiud commented:
> Do you have an approximate notion of how expensive it is to leave this on?

Nope. This is equivalent to the greedy search for transfers. Scales as N^2 mpi procs

> Is there a risk of this biting people in moving mesh simulations? Like, they do a few timesteps, they're comfortable that they haven't got any transfer ambiguity, then they start a whole run and the next thing you know timestep 200 dies because a deforming mesh put a node into the wrong place.

Depends what you mean by biting. It s just a warning so it wont crash their simulation. 
If they turn it off, then yes they can be hit by getting a FP round-off determined value. But then again they turned it off

GiudGiud commented:
Should be mergable independent from the two other PRs btw

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1529363/) on d77f93e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>587ef5</th>
      <th colspan=4>#24196 d77f93</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.21%</td>
      <td>85.21%</td>
      <td>-0.00%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87702</td>
      <td>87706</td>
      <td>+4</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15217</td>
      <td>15218</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24196/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24196/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24196/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 66.67% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Seems there's some test failures to investigate in sweeping. Probably more value conflicts but need to make sure

GiudGiud committed:
Fix resolving of value conflict

GiudGiud commented:
Should be green now.

GiudGiud commented:
These have got to be false positives. I don’t think I was that bad at designing my tests..

GiudGiud committed:
Rename GeneralField NearestNode to GeneralField nearest location
it doesnt only use nodes

GiudGiud committed:
Deprecate the previous generation of transfers
Because:
- they dont offer as much parameter checking
notably for higher order variables
- they dont have options to see if the transfer is well posed
- we cant be supporting 2 objects doing the same thing but slightly differently
refs #24226

GiudGiud commented:
ok found the problem. It s a false negative that appears when:
- there is an indetermination in what is being sent by a process (think rank 3 holds 2 points that are equidistant from any target point and the closest in rank 3 to that point)
- that value is not actually the "nearest" so it doesnt matter (those 2 points are far from the target point, or simply not as close a point in rank 1)

GiudGiud commented:
of course the problem only shows itself in nearest node, the most awful one. The other ones dont allow for this nonsense

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1651652/) on 10438c9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24196/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24196/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c7d87531e5eba70adef4e54fcf958d6a487b3d7c`

GiudGiud committed:
Resolve local conflicts by communicating them to all ranks
- when looking for a value for a transfer, it could be that two source values were just as close, which is considered a value conflict for the transfer
However if a third value was closer on another process, the conflict is not materialized
- be consistent about "is_nodal" not being "is_lagrange" even if is_nodal is an API in need of a rework
refs #24226

GiudGiud committed:
More test fixups
Turning on the search value conflict is effectively triggering a greedy search, which unfortunately
changes the results for only some processor numbers (which is incorrect). This is likely a case of indetermination that is not
picked up by the diagnostics

The tests concerned are all in this commit

GiudGiud commented:
Last commit explains some testing issues. I dont think I m breaking things with this PR, but likely displaying an area of improvement for nearest node as evidenced by some parallel diffs with greedy search

roystgnr commented on framework/include/transfers/MultiAppGeneralFieldTransfer.h:
```diff
@@ -320,15 +320,41 @@ class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
+      const unsigned int var_index,
       const DofobjectToInterpValVec & dofobject_to_valsvec,
       const InterpCaches & distance_caches,
-      std::vector<std::tuple<unsigned int, dof_id_type, Point, Real>> conflicts_vec);
```
This was an atrocity.  We were accidentally passing by value before, so discarding changes entirely?

roystgnr replied:
The more I think about it ... does this need to be a separate parameter, even?  We only ever call it with `this->_local_conflicts`, right?  Likewise with `this->_received_conflicts` in that version above?  And this is a non-const non-static member function so we already have access to those that way.

GiudGiud replied:
I can remove it from the parameters. Initially I was trying to use the same routine both local and received conflicts but that didn’t last

roystgnr replied:
I'm fine with it as-is, if you don't want to change it.  Almost done with the rest of the review, and so far I haven't seen anything that actually *needs* a change.

roystgnr commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -743,26 +744,25 @@ MultiAppGeneralFieldTransfer::examineValueConflicts(
-    // Extract variable info
+    // Extract target variable info
     auto & es = getEquationSystem(*_to_problems[problem_id], _displaced_target_mesh);
     System * to_sys = find_sys(es, var_name);
```
Would it be hard to just save this `System *`, the same way we save the var name now?  Going from a perfectly good index to a string makes me worry about performance (string comparisons don't belong in inner loops) and worry more about robustness (now that we can support multiple systems it's easy to imagine use cases in which we'd want them to have overlapping variable names...)

GiudGiud replied:
@friedmud brought the same concern (in the first general field PR) as it's done all over with the names (in fact saving the system does not help, you need to save the variable number as that's retrieved with string comparisons too)

and I profiled and it was irrelevant

overlapping variable names sounds like a bad idea

GiudGiud committed:
Address Roy's review:
- remove an argument to the value conflict examination that is trivially the class attribute

moosebuild commented:
All [jobs](https://civet.inl.gov/event/136941/) on 09cafa5 : invalidated by @GiudGiud

fluke

GiudGiud commented:
@roystgnr

cticenhour commented on modules/doc/content/newsletter/2023/2023_02.md:
```diff
@@ -9,7 +9,7 @@ based on libMesh's [generic projector](https://mooseframework.inl.gov/docs/doxyg
 
 - [MultiAppGeneralFieldShapeEvaluationTransfer.md]
-- [MultiAppGeneralFieldNearestNodeTransfer.md]
+- [MultiAppGeneralFieldNearestLocationTransfer.md]
```
Thanks for updating these old newsletters. Also thanks for all the transitions to parent/child in the in-file doco and elsewhere.

GiudGiud replied:
you're welcome!

GiudGiud commented:
so green

cticenhour commented:
@roystgnr any more comments/review on this PR? I am happy to finish off the review, but I am not as familiar with the current transfers system.

roystgnr commented:
Man, that was so much easier than 25087.  :-)  Looks good to me now!

GiudGiud commented:
;)

thanks for the review! Last change to transfers hopefully for a while! Hopefully no other false positives to address with this new check

aprilnovak commented:
FYI this is causing problems with moving mesh cases in Cardinal (see: https://github.com/neams-th-coe/cardinal/pull/788). Sending displacements into Nek, our mesh elements are now giving negative Jacobians. Sorry for being so late to the discussion here, it took some time to do the git bisection search to identify the cause, and we only update MOOSE ~once a month.

<img width="1046" alt="Screen Shot 2023-10-07 at 10 16 38 AM" src="https://github.com/idaholab/moose/assets/17039662/9a6984f8-c357-4082-85c4-f6d78333b891">

My two cents is that this flag would be better defaulting to "off," if it takes additional time to evaluate & has potential to break moving mesh cases.

GiudGiud commented:
We pushed a fix recently (last week)

Which commit are you using?

aprilnovak commented:
Thanks for the tip! I was using a commit just a tad older than that. The hotfix solves our issue, thanks so much.
